transports/quic/src/socket.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! I/O for libp2p-quic.
+//!
+//! This provides a central location for socket I/O, and logs all incoming and outgoing packets.
+
+use async_std::net::UdpSocket;
+use log::{trace, warn};
+use quinn_proto::Transmit;
+use std::{future::Future, io::Result, task::Context, task::Poll};
+
+/// A pending packet for libp2p-quic
+#[derive(Debug, Default)]
+pub(crate) struct Pending {
+    pending: Option<Transmit>,
+}
+
+/// A socket wrapper for libp2p-quic
+#[derive(Debug)]
+pub(crate) struct Socket {
+    socket: UdpSocket,
+}
+
+impl Socket {
+    /// Transmit a packet if possible, with appropriate logging.
+    ///
+    /// We ignore I/O errors.  If a packet cannot be sent, we assume it is a transient condition
+    /// and drop it.  If it is not, the connection will eventually time out.  This provides a very
+    /// high degree of robustness.  Connections will transparently resume after a transient network
+    /// outage, and problems that are specific to one peer will not effect other peers.
+    pub(crate) fn poll_send_to(&self, cx: &mut Context<'_>, packet: &Transmit) -> Poll<Result<()>> {
+        match {
+            let fut = self.socket.send_to(&packet.contents, &packet.destination);
```

Comment:
Is it safe to drop the future in case it returns `Pending`? We had a similar discussion in https://github.com/libp2p/rust-libp2p/pull/1247#discussion_r325751157. As of today [`send_to` implementation](https://docs.rs/async-std/0.99.5/src/async_std/net/udp/mod.rs.html#152-176) seems stateless, but that might change without a breaking change, right?

Reply:
The async-std developers [suggested this].  I don’t like it any more than you do, but I was not able to find a better alternative.

[suggested this]: /async-rs/async-std/issues/634#issuecomment-573861865